runtime/src/lib.rs

```diff
@@ -545,6 +545,45 @@ impl frame_system::offchain::CreateTransaction<Runtime, UncheckedExtrinsic> for
 	}
 }
 
+/// `DelegatedDispatchVerifier` helpers
+impl Runtime {
+	/// Check that the caller is the `holder` of the doughnut
+	fn check_caller(
+		caller: &<Runtime as frame_system::Trait>::AccountId,
+		doughnut: &<Runtime as frame_system::Trait>::Doughnut,
+	) -> Result<(), &'static str> {
+		if caller.clone() != doughnut.holder() {
+			Err("Invalid doughnut holder")
+		} else {
+			Ok(())
+		}
+	}
+
+	/// Checks if the doughnut holds a `cennznut` and if so, it returns the decoded `cennznut`
+	fn get_cennznut(doughnut: &<Runtime as frame_system::Trait>::Doughnut) -> Result<CENNZnut, &'static str> {
+		let mut domain = doughnut
+			.get_domain(<Runtime as additional_traits::DelegatedDispatchVerifier>::DOMAIN)
+			.ok_or("CENNZnut does not grant permission for cennznet domain")?;
+		let cennznut: CENNZnut = Decode::decode(&mut domain).map_err(|_| "Bad CENNZnut encoding")?;
+		Ok(cennznut)
+	}
+
+	/// Verify that the smart contract being called is permissioned in the cennznut
+	fn check_contract(
+		contract_address: &<Runtime as frame_system::Trait>::AccountId,
+		cennznut: CENNZnut,
+	) -> Result<(), &'static str> {
+		let address: [u8; 32] = contract_address.clone().into();
+		match cennznut.validate_contract_call(&address) {
+			Ok(r) => Ok(r),
+			Err(ValidationErr::NoPermission(ContractDomain::Contract)) => {
```

Comment:
Any reason for checking two different conditions but return same err?

Reply:
That's a good point - could just make it:
`_ => Err("CENNZnut does not grant permission for contract")`